### PR TITLE
h3: generate event for stream close

### PR DESCRIPTION
**Pull request by LPardue**

### CODE REVIEW

File: src/h3/mod.rs [hunk 1]
```diff
@@ -456,6 +464,9 @@ pub enum Event {
 
     /// Data was received.
     Data(Vec<u8>),
+
+    // Stream was closed,
+    StreamClosed(),
```
Comment (ghedo) [inline src/h3/mod.rs]:
```suggestion
    Finished,
```

So a couple of things: 1. I don't think you need `()` since you are not actually wrapping another type (this also means that matching this event would be somewhat more readable) 2. `Finished` seems like a better name (and shorter), but YMMV.

File: src/h3/ffi.rs [hunk 2]
```diff
@@ -121,6 +123,8 @@ pub extern fn quiche_h3_event_for_each_header(
             },
 
         h3::Event::Data { .. } => unreachable!(),
+
+        h3::Event::StreamClosed { .. } => unreachable!(),
```
Comment (ghedo) [inline src/h3/ffi.rs]: You could probably just replace this and the previous case with `_ => unreachable!()`.

File: src/h3/ffi.rs [hunk 3]
```diff
@@ -138,6 +142,19 @@ pub extern fn quiche_h3_event_data(
 
             data.len()
         },
+
+        h3::Event::StreamClosed { .. } => unreachable!(),
```
Comment (ghedo) [inline src/h3/ffi.rs]: Same, re: using `_ => unreachable!()`.

File: src/h3/ffi.rs [hunk 4]
```diff
@@ -93,6 +93,8 @@ pub extern fn quiche_h3_event_type(ev: &h3::Event) -> u32 {
         h3::Event::Headers { .. } => 0,
 
         h3::Event::Data { .. } => 1,
+
+        h3::Event::StreamClosed { .. } => 2,
```
Comment (ghedo) [inline src/h3/ffi.rs]: You also need to add this to the `quiche_h3_event_type` enum in `quiche.h`.

File: src/h3/ffi.rs [hunk 5]
```diff
@@ -138,6 +142,19 @@ pub extern fn quiche_h3_event_data(
 
             data.len()
         },
+
+        h3::Event::StreamClosed { .. } => unreachable!(),
+    }
+}
+
+#[no_mangle]
+pub extern fn quiche_h3_event_stream_closed(ev: &h3::Event) {
```
Comment (ghedo) [inline src/h3/ffi.rs]: I don't think this is needed? It also doesn't seem to do anything.

**Comment by ghedo** [comment]
Oh also need to update https://docs.quic.tech/quiche/h3/index.html#detecting-end-of-stream

**Comment by LPardue** [comment]
Thanks. I'll address your comments tomorrow

**Comment by LPardue** [comment]
The latest changes are almost there but have an edge case that needs handling.

I also plan to rework unit tests slightly.

File: src/h3/mod.rs [hunk 6]
```diff
@@ -1371,8 +1532,513 @@ mod tests {
...
+                &mut s.pipe.client,
+                s.client.peer_qpack_streams.decoder_stream_id.unwrap(),
+                b"Uh oh",
+                true
+            ),
+            Err(Error::WrongStream)
+        );
+
+        // server-side streams
+        assert_eq!(
```
Comment (ghedo) [inline src/h3/mod.rs]: Do we really need to duplicate the same tests for both sides? I'd expect that the same code handles both cases anyway.

File: src/h3/mod.rs [hunk 7]
```diff
@@ -952,6 +1009,15 @@ impl Connection {
 
     /// Opens QPACK encoder and decoder streams, if not already opened.
     fn open_qpack_streams(&mut self, conn: &mut super::Connection) -> Result<()> {
+        self.open_qpack_encoder_stream(conn)?;
```
Comment (ghedo) [inline src/h3/mod.rs]: I guess it's kind of unrelated, but couldn't we just have a single "open_uni_stream()" that takes the stream type and returns the stream ID, so then we could do e.g. `self.local_qpack_streams.encoder_stream_id = open_uni_stream(QPACK_ENCODER_STREAM_TYPE_ID);` or something like that?

File: src/h3/mod.rs [hunk 8]
```diff
@@ -1309,10 +1411,69 @@ mod tests {
 
             self.pipe.advance(buf).ok();
 
+            // server streams
```
Comment (ghedo) [inline src/h3/mod.rs]:
```suggestion
            // Server streams.
```

File: src/h3/mod.rs [hunk 9]
```diff
@@ -1294,13 +1389,20 @@ mod tests {
             })
         }
 
+        /// Do the HTTP/3 handshake so both ends are in sane initial state
         pub fn handshake(&mut self, buf: &mut [u8]) -> Result<()> {
             self.pipe.handshake(buf)?;
 
+            // client streams
```
Comment (ghedo) [inline src/h3/mod.rs]:
```suggestion
            // Client streams.
```

File: src/h3/mod.rs [hunk 10]
```diff
@@ -1309,10 +1411,69 @@ mod tests {
...
+            self.pipe.advance(buf).ok();
+
+            if self.pipe.server.grease {
+                self.server.open_grease_stream(&mut self.pipe.server)?;
+            }
+
+            self.pipe.advance(buf).ok();
+
+            loop {
+                match self.client.poll(&mut self.pipe.client) {
```
Comment (ghedo) [inline src/h3/mod.rs]: I think you could just turn this into something like:
```rust
while self.client.poll(&mut self.pipe.client).is_ok() {
  // Do nothing.
}
```

Same for the one below.

File: src/h3/mod.rs [hunk 11]
```diff
@@ -1309,10 +1411,69 @@ mod tests {
...
+                    Err(_) => (),
+                }
+            }
+
             Ok(())
         }
+
+        /// Calls poll on identified role and asserts the returned event matches
+        /// the expected event.
+        pub fn poll_assert(
```
[the remark this answers is not in the document]
Comment (LPardue) [reply src/h3/mod.rs]: This function has some pros and cons:

pro - it avoids long assert lines that get wrapped by cargo fmt, so reduces line count
pro - it avoids code goofs where you poll a H3 connection with the wrong socket

con - any assert failures no longer happen in the test, so its very difficult to quickly diagnose why a test failed
con - Role enum

Suggest we keep poll_assert for now and refactor the way it works to be a macro that keeps pros but removes cons.

File: src/h3/mod.rs [hunk 12]
```diff
@@ -657,6 +682,20 @@ impl Connection {
         // Return how many bytes were written, excluding the frame header.
         let written = conn.stream_send(stream_id, body, fin)?;
 
+        // tidy up streams
+        if !fin {
+            return Ok(written);
+        }
+
+        if self
```
Comment (RReverser) [inline src/h3/mod.rs]: Try instead something like:

```rust
match self.streams.entry(stream_id) {
  Entry::Occupied(entry) if entry.get().peer_fin() => {
    entry.remove();
  }
  _ => {}
}
```

File: src/h3/mod.rs (hunk 7 above)
Comment (LPardue) [reply src/h3/mod.rs]: lol yes that is a much better suggestion

File: src/h3/mod.rs (hunk 6 above)
Comment (LPardue) [reply src/h3/mod.rs]: Change the test to focus on server.

If client-side tests are different/needed, we can add a different test later.

File: src/h3/mod.rs [hunk 13]
```diff
@@ -657,6 +682,20 @@ impl Connection {
         // Return how many bytes were written, excluding the frame header.
         let written = conn.stream_send(stream_id, body, fin)?;
 
+        // tidy up streams
+        if !fin {
```
[the remark this answers is not in the document]
Comment (LPardue) [reply src/h3/mod.rs]: done

File: src/h3/mod.rs (hunk 12 above)
Comment (RReverser) [reply src/h3/mod.rs]: Oh? What failed?

Comment (LPardue) [reply src/h3/mod.rs]: Compiler complained that Entry was unknown. A tried a few things but gave up.

Comment (RReverser) [reply src/h3/mod.rs]: Unknown as in, you just need to `use ...` it to get in scope first?
